### PR TITLE
Added comlink and a sqlite client that encapsulates the communication…

### PR DESCRIPTION
**Pull request by etiennenoel**
… with the worker.

### CODE REVIEW

**Comment by etiennenoel** [comment]
@tomayac if you can review the code and let me know what you think. I've updated the demo. I'm not sure about hardcoding the link to the comlink cdn but the alternative is to package it somehow and I don't think that's better either.

Also, I will update the readme.md once (and if) we are satistified with the code.

**Comment by tomayac** [comment]
Thanks, this is great, and allows people to either use this convenience layer, or ignore it and use the traditional approach. 

Regarding the Comlink dependency, just `npm install` it, and bundlers will be smart about only including it if it’s actually needed. For the demo, hard coding a CDN concretely is fine, but for the `src/` code you want to avoid hard coded dependencies.

File: demo/index.html [hunk 1]
```diff
@@ -5,12 +5,15 @@
     <title>SQLite Wasm Demo</title>
     <script type="module" src="script.js"></script>
     <script type="module" src="main-thread.js"></script>
+    <script type="module" src="sqlite.client.js"></script>
```
Comment (tomayac) [inline demo/index.html:8]: Can you stick to the `-` naming convention?

File: demo/index.html [hunk 2]
```diff
@@ -5,12 +5,15 @@
...
   </head>
   <body>
     <h1>SQLite Wasm Demo</h1>
     <h2>Main thread</h2>
     <div class="main-thread"></div>
     <h2>Worker</h2>
     <div class="worker"></div>
+    <h2>Built-in Sqlite Client</h2>
```
Comment (tomayac) [inline demo/index.html:16]:
```suggestion
    <h2>Built-in SQLite Client</h2>
```

Comment (tomayac) [reply demo/index.html:16]: Actually, maybe hint at what this does under the hood. Maybe get inspiration from Comlink’s README.

File: demo/sqlite.client.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+for (let i = 20; i <= 25; ++i) {
+  await sqliteClient.executeSql('INSERT INTO t(a,b) VALUES (?,?)', [i, i * 2]);
+}
+
+const rows = await sqliteClient.executeSql('SELECT a FROM t ORDER BY a LIMIT 3');
+
+console.log(rows);
+
+document.getElementById('sqlite-client').innerHTML =
```
Comment (tomayac) [inline demo/sqlite.client.js:17]: Can you use a template string here and nest it properly?

File: package.json [hunk 4]
```diff
@@ -58,5 +58,6 @@
     "prettier": "^2.8.8",
     "publint": "^0.1.12",
     "shx": "^0.3.4"
-  }
+  },
+  "dependencies": {}
```
Comment (tomayac) [inline package.json:62]: This looks like a leftover. `npm install`ing Comlink should fix it.

File: src/sqlite.client.mjs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  async executeSql(sqlStatement, bindParameters = []) {
+    if (typeof sqlStatement !== 'string') {
+      return error(
+        "The 'sqlStatement' parameter passed to the 'executeSql' method of the SqliteClient must be of type 'string'. Instead, you passed: '" +
+          typeof sqlStatement +
+          "'",
+      );
+    }
+
+    if (Array.isArray(bindParameters) === false) {
```
Comment (tomayac) [inline src/sqlite.client.mjs:53]: `if (!Array.isArray(…))`

File: src/sqlite.worker.mjs [hunk 6]
```diff
@@ -0,0 +1,39 @@
+import * as Comlink from "https://unpkg.com/comlink/dist/esm/comlink.mjs";
+import {default as sqlite3InitModule} from "../index.mjs"
+
+const log = (...args) => console.log(...args);
+const error = (...args) => console.error(...args);
+
+
```
Comment (tomayac) [inline src/sqlite.worker.mjs:7]: Can you leave at max 1 free line throughout the file?

**Review by tomayac — COMMENTED** [review]
Getting there. I like where this is headed. Thanks for working on this!

File: package.json (hunk 4 above)
Comment (etiennenoel) [reply package.json:62]: I'm actually not depending on it as I kind of need to use the cdn else we would need to package it somehow. Will remove this

Comment (tomayac) [reply package.json:62]: See my comment elsewhere. We want to depend on Comlink. Exception is the demo.

Comment (etiennenoel) [reply package.json:62]: Oups I had not seen your other comment. Will take a complete look tomorrow.

**Comment by etiennenoel** [comment]
@tomayac I understand not including dependencies in the `/src` folder. In the case of the `sqlite-worker.mjs`, if I include a normal npm dependency as `import {comlink} from "comlink"`, I will have to bundle it somehow to make the demo work.

What's the easiest way I should do this you think? (trying to avoid including too many dependencies in here).

**Comment by etiennenoel** [comment]
@tomayac I understand not including dependencies in the `/src` folder. In the case of the `sqlite-worker.mjs`, if I include a normal npm dependency as `import {comlink} from "comlink"`, I will have to bundle it somehow to make the demo work.

What's the easiest way I should do this you think? (trying to avoid including too many dependencies in here).

File: demo/index.html (hunk 1 above)
Comment (etiennenoel) [reply demo/index.html]: Done.

File: package.json (hunk 4 above)
Comment (etiennenoel) [reply package.json]: Done.

File: src/sqlite.worker.mjs (hunk 6 above)
Comment (etiennenoel) [reply src/sqlite.worker.mjs]: Done.

**Comment by schickling** [comment]
Sorry for the late response but as a early user (and given I've built a [predecessor of this project](https://github.com/overtone-app/sqlite-wasm-esm)), I'd like to voice a concern in regards to adding `comlink` as a dependency:

I've used `comlink` quite a bit in previous projects, and while it certainly provides some convenience when working with workers, it also adds a considerable amount of performance (and also bundle size) overhead. Unfortunately I don't have any kind of benchmark numbers available right now, but it was significant enough for my use cases, that I've removed `comlink` where it became a bottleneck. (Btw still kudos to the folks who've created `comlink` as it definitely helps when getting started with workers in the web.)

Given `sqlite-wasm` is meant to be used (to my knowledge) as a fairly low-level wrapper around SQLite's WASM build, I feel pretty strongly about avoid further dependencies and performance overhead however possible.


Side comment: the merged source file is currently importing from `unpkg` and not from the locally installed dependency:

CleanShot 2023-07-04 at 12 15 39@2x

**Comment by tomayac** [comment]
Thanks for chiming in. I’ve made changes to the finally released version that imports locally, and not from the CDN. Comlink is used as a convenience layer for those who like it, the original way of using the library is still available unchanged.

**Comment by schickling** [comment]
> Comlink is used as a convenience layer for those who like it, the original way of using the library is still available unchanged.

That's good to know. Would you be open to making `comlink` an optional dependency to keep the core of the package minimal? Another approach could be to create a separate package with the new convenience layer.

**Comment by tomayac** [comment]
We can make Comlink an optional dependency, sure. Can you please open a quick PR? I’m OoO until the 17th, but happy to merge it as soon as I return.

**Comment by etiennenoel** [comment]
I've created one here: https://github.com/tomayac/sqlite-wasm/pull/20

**Pull request merged**